token/program-2022/src/extension/confidential_transfer/processor.rs

```diff
@@ -279,94 +279,74 @@ fn process_deposit(
         return Err(TokenError::MintDecimalsMismatch.into());
     }
 
-    // Process source account
-    {
-        check_program_account(token_account_info.owner)?;
-        let token_account_data = &mut token_account_info.data.borrow_mut();
-        let mut token_account = StateWithExtensionsMut::<Account>::unpack(token_account_data)?;
-
-        Processor::validate_owner(
-            program_id,
-            &token_account.base.owner,
-            authority_info,
-            authority_info_data_len,
-            account_info_iter.as_slice(),
-        )?;
-
-        if token_account.base.is_frozen() {
-            return Err(TokenError::AccountFrozen.into());
-        }
+    check_program_account(token_account_info.owner)?;
+    let token_account_data = &mut token_account_info.data.borrow_mut();
+    let mut token_account = StateWithExtensionsMut::<Account>::unpack(token_account_data)?;
 
-        if token_account.base.mint != *mint_info.key {
-            return Err(TokenError::MintMismatch.into());
-        }
+    Processor::validate_owner(
+        program_id,
+        &token_account.base.owner,
+        authority_info,
+        authority_info_data_len,
+        account_info_iter.as_slice(),
+    )?;
 
-        // Wrapped SOL deposits are not supported because lamports cannot be vanished.
-        assert!(!token_account.base.is_native());
-        token_account.base.amount = token_account
-            .base
-            .amount
-            .checked_sub(amount)
-            .ok_or(TokenError::Overflow)?;
+    if token_account.base.is_frozen() {
+        return Err(TokenError::AccountFrozen.into());
+    }
 
-        token_account.pack_base();
+    if token_account.base.mint != *mint_info.key {
+        return Err(TokenError::MintMismatch.into());
     }
 
-    //
-    // Finished with the source token account at this point. Drop all references to it to avoid a
-    // double borrow if the source and destination accounts are the same
-    //
+    // Wrapped SOL deposits are not supported because lamports cannot be vanished.
+    assert!(!token_account.base.is_native());
 
-    // Process destination account
-    {
-        check_program_account(destination_token_account_info.owner)?;
-        let destination_token_account_data = &mut destination_token_account_info.data.borrow_mut();
-        let mut destination_token_account =
-            StateWithExtensionsMut::<Account>::unpack(destination_token_account_data)?;
+    // A deposit amount must be a 48-bit number
+    if amount >> MAXIMUM_DEPOSIT_TRANSFER_AMOUNT_BIT_LENGTH > 0 {
+        return Err(TokenError::MaximumDepositAmountExceeded.into());
+    }
```

Comment:
is this sufficient to remediate HAL-03? legitimate question, i dont understand a lot of what happens in this extension to be honest

their comments in their report, and the documentation they linked [here](https://spl.solana.com/confidential-token/deep-dive/overview#dealing-with-discrete-log), seem to indicate balances must never exceed u48::MAX. but the math down below with `pending_balance_lo` and `pending_balance_hi` makes it seem like values up to u64::MAX are expected?

also, their comments state that HAL-03 breaks the decryption primitive via an "overflow," but the protocol overview doc claims a 48-bit limit only exists for performance reasons? i am doubly unsure about all of this after reading the zk token paper because it describes splitting the balance into two 32-bit numbers for performance reasons, so im not sure what the 16/48 split here is meant to be 

anyway, you dont need to explain the cryptography to me, i lack the background to appreciate it. assuming the HAL-03 report is correct, and this block is the only new safeguard against it, here are my attacks:
* i break my own account by depositing u48::MAX and then depositing again
* i break my own account by depositing u48::MAX and rolling that pending balance into a nonzero active balance. equivalently, i transfer to someone else and attempt to convince them to roll the pending into their nonzero active
* i break someone elses account by depositing u48::MAX into my own account, transfering to them, depositing again, transfering again. skip the second cycle if their balance was nonzero

do these work? or am i missing something?

Reply:
You are understanding things correctly. All of these are valid points. Restricting the amounts to 48-bit values was added on after the paper was written, so it is a little confusing. I discuss this a little bit in the [spl doc](https://spl.solana.com/confidential-token/deep-dive/encryption), which is more up-to-date than the paper.

All the attacks that you mentioned are valid. I do want to mention though that if the funds do overflow, it is not that they are lost forever. It is just that decryption become noticably slower. If the funds overflow, then (1) one can spend more time decrypting it or (2) it can also look up individual transaction data that sent tokens to the confidential account and then decrypt the associated amounts individually (the amounts associated with the instructions are guaranteed to be 48-bits by zkp).

I can comment on the attacks in order:
- There is no mechanism to protect against this attack. If this happens by accident, then the owner can just do the costly decryption or decrypts the instruction data individually.
- Only pending balance is not decryptable. Once pending balance is rolled over to available balance, then the balances are decryptable because available balance is encrypted as decryptable balance.
- There is no mechanism to protect against this attack. If this happens, then the receiver has to do the costly decryption (after disabling incoming transfers if needed) or decrypt the individual instruction data. This is a more serious attack than the above two because a malicious sender can genuinely impact another user's accounts. Unfortunately, no mechanism is possible to prevent this without significantly impacting performance. Practically speaking, it would be a very costly attack for the malicious sender only to overflow an account which the receiver can still recover the funds for (albeit in a more costly way).

Reply:
ahhh i see. their line:
>Doing so will cause the
 Elgamal encryption addition function to malfunction/overflow and the
 users available_balance will not be decryptable thus leading to a loss
 of funds.

led me to believe this was more serious than merely slowing it down